libvmi/driver/xen/xen.c

```diff
@@ -280,6 +280,70 @@ status_t xen_get_name_from_domainid(
 }
 #endif
 
+#ifndef HAVE_LIBXENSTORE
+uint64_t xen_get_domainid_from_uuid(
+    vmi_instance_t UNUSED(vmi),
+    const char* UNUSED(uuid))
+{
+    return VMI_FAILURE;
+}
+#else
+uint64_t xen_get_domainid_from_uuid(
+    vmi_instance_t vmi,
+    const char* uuid)
+{
+    if (uuid == NULL) {
+        return VMI_INVALID_DOMID;
+    }
+
+    xen_instance_t *xen = xen_get_instance(vmi);
+    char **domains = NULL;
+    unsigned int size = 0, i = 0;
+    xs_transaction_t xth = XBT_NULL;
+    uint64_t domainid = VMI_INVALID_DOMID;
+
+    struct xs_handle *xsh = xen->libxsw.xs_open(0);
+
+    if (!xsh)
+        goto _bail;
+
+    domains = xen->libxsw.xs_directory(xsh, xth, "/local/domain", &size);
+    for (i = 0; i < size; ++i) {
+        /* read in ID */
+        char *idStr = domains[i];
+        char *tmp = g_strconcat("/local/domain/", idStr, "/vm", NULL);
+        char *path = xen->libxsw.xs_read(xsh, xth, tmp, NULL);
+
+        if (path && path[0] != '\0') {
+            tmp = g_strconcat(path, "/uuid", NULL);
+
+            char *uuidCandidate = xen->libxsw.xs_read(xsh, xth, tmp, NULL);
+
+            /* if uuid matches, then return number */
+            if (uuidCandidate != NULL &&
+                    strncmp(uuid, uuidCandidate, 100) == 0) {
+                domainid = strtoull(idStr, NULL, 0);
+                free(uuidCandidate);
+                free(path);
+                break;
+            }
+            /* free memory as we go */
+            if (path)
```

Comment:
Just use `g_free` that is already null-safe

Reply:
free() does nothing if the pointer is null. I can drop the if() and if the g_free() is a must I can change it.

Reply:
On most C runtimes that is true but for compatibility reasons `g_free` is nicer.